src/evo/assetlocktx.cpp

```diff
@@ -0,0 +1,249 @@
+}
+
+uint8_t CAssetLockPayload::getVersion() const
+{
+    return nVersion;
+}
+
+const std::vector<CTxOut>& CAssetLockPayload::getCreditOutputs() const
+{
+    return creditOutputs;
+}
+
+/*
+   Asset Unlock Transaction (withdrawals)
+   */
+
+const std::string ASSETUNLOCK_REQUESTID_PREFIX = "plwdtx";
+
+bool CAssetUnlockPayload::VerifySig(const uint256& msgHash, const CBlockIndex* pindexTip, TxValidationState& state) const
+{
+    // That quourm hash must be active at `requestHeight`,
+    // and at the quorumHash must be active in either the current or previous quorum cycle
+    // and the sig must validate against that specific quorumHash.
+
+    Consensus::LLMQType llmqType = Params().GetConsensus().llmqTypeAssetLocks;
+
+    // We check at most 2 quorums
+    const auto quorums = llmq::quorumManager->ScanQuorums(llmqType, pindexTip, 2);
+    bool isActive = std::any_of(quorums.begin(), quorums.end(), [&](const auto &q) { return q->qc->quorumHash == quorumHash; });
+
+    if (!isActive) {
+        return state.Invalid(TxValidationResult::TX_CONSENSUS, "bad-assetunlock-not-active-quorum");
+    }
+
+    if (pindexTip->nHeight < requestedHeight || pindexTip->nHeight >= getHeightToExpiry()) {
+        LogPrintf("Asset unlock tx %d with requested height %d could not be accepted on height: %d\n",
+                index, requestedHeight, pindexTip->nHeight);
+        return state.Invalid(TxValidationResult::TX_CONSENSUS, "bad-assetunlock-too-late");
+    }
+
+    const auto quorum = llmq::quorumManager->GetQuorum(llmqType, quorumHash);
+    assert(quorum);
+
+    const uint256 requestId = ::SerializeHash(std::make_pair(ASSETUNLOCK_REQUESTID_PREFIX, index));
+
+    const uint256 signHash = llmq::utils::BuildSignHash(llmqType, quorum->qc->quorumHash, requestId, msgHash);
+    if (quorumSig.VerifyInsecure(quorum->qc->quorumPublicKey, signHash)) {
+        return true;
+    }
+
+    return state.Invalid(TxValidationResult::TX_CONSENSUS, "bad-assetunlock-not-verified");
+}
+
+bool CheckAssetUnlockTx(const CTransaction& tx, const CBlockIndex* pindexPrev, const CCreditPool& creditPool, TxValidationState& state)
+{
+    if (tx.nType != TRANSACTION_ASSET_UNLOCK) {
+        return state.Invalid(TxValidationResult::TX_BAD_SPECIAL, "bad-assetunlocktx-type");
+    }
+
+    if (!tx.vin.empty()) {
+        return state.Invalid(TxValidationResult::TX_BAD_SPECIAL, "bad-assetunlocktx-have-input");
+    }
+
+    if (tx.vout.size() > CAssetUnlockPayload::MAXIMUM_WITHDRAWALS) {
+        return state.Invalid(TxValidationResult::TX_BAD_SPECIAL, "bad-assetunlocktx-too-many-outs");
+    }
+
+    CAssetUnlockPayload assetUnlockTx;
+    if (!GetTxPayload(tx, assetUnlockTx)) {
+        return state.Invalid(TxValidationResult::TX_BAD_SPECIAL, "bad-assetunlocktx-payload");
+    }
+
+    if (assetUnlockTx.getVersion() == 0 || assetUnlockTx.getVersion() > CAssetUnlockPayload::CURRENT_VERSION) {
```

Comment:
please don't use public getters like this when we have access to the actual variable; this isn't java :D I'd prefer to have minimal getter usage

Reply:
Okay, nevermind; I see we don't have access here... I don't love the getter usage but I guess that's fine

Reply:
I can't make members constant, because READWRITE macros for serialization/de-serialization.

Because the members of `CAsset{Lock,Unlock}Payload` are not constant, better to have them private -> need to have getters. If there were other way to do serialization/deserialization without bunch of boiler code I'd like to get rid of these  getters too